internal/job/job.go

```diff
@@ -0,0 +1,223 @@
+package job
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"sync/atomic"
+
+	"github.com/joshuarubin/teleport-job-worker/internal/config"
+	"github.com/joshuarubin/teleport-job-worker/internal/safebuffer"
+	"github.com/joshuarubin/teleport-job-worker/internal/user"
+)
+
+// Job represents a system process
+type Job struct {
+	cfg    *config.Config
+	id     ID
+	userID user.ID
+	cmd    *exec.Cmd
+	buf    *safebuffer.Buffer
+
+	startOnce sync.Once
+
+	stopOnce sync.Once
+	stopErr  atomic.Value
+
+	status atomic.Uint32
+
+	done   chan struct{}
+	cmdErr error // only safe to read after done has closed
+}
+
+// New creates, but does not start a new job
+func New(
+	ctx context.Context,
+	cfg *config.Config,
+	userID user.ID,
+	command string,
+	args []string,
+	env []string,
+) (*Job, error) {
+	id, err := NewID()
+	if err != nil {
+		return nil, err
+	}
+
+	done := make(chan struct{})
+
+	j := Job{
+		cfg:    cfg,
+		id:     id,
+		userID: userID,
+		done:   done,
+		cmd:    exec.CommandContext(ctx, command, args...),
+	}
+
+	j.buf = safebuffer.New(ctx, j.done)
+	j.cmd.Stdout = j.buf
+	j.cmd.Stderr = j.buf
+
+	j.cmd.SysProcAttr = sysProcAttr()
+
+	j.cmd.Env = os.Environ()
+	j.cmd.Env = append(j.cmd.Env, env...)
+	j.setStatus(StatusNotStarted)
+
+	return &j, nil
+}
+
+var ErrAlreadyStarted = errors.New("already started")
+
+// Start the job process. If Start() is called more than once ErrAlreadyStarted
+// will be returned.
+func (j *Job) Start() error {
+	var err error
+	var started bool
+	j.startOnce.Do(func() {
+		j.setStatus(StatusRunning)
+		err = j.cmd.Start()
+		started = true
+
+		go func() {
+			defer func() {
+				j.setStatus(StatusComplete)
+				close(j.done)
+			}()
+			j.cmdErr = j.cmd.Wait()
+		}()
+	})
+
+	if !started {
+		return ErrAlreadyStarted
+	}
+
+	return err
+}
+
+// ID returns the job ID
+func (j *Job) ID() ID {
+	return j.id
+}
+
+// UserID returns the job ID
+func (j *Job) UserID() user.ID {
+	return j.userID
+}
+
+// NewOutputReader returns an io.ReadCloser that can be used to stream the
+// output of the job from the time it started. It is the caller's responsibility
+// to close it to free allocated resources.
+func (j *Job) NewOutputReader() io.ReadCloser {
+	return j.buf.NewReader()
+}
+
+// Done returns a channel that will be closed when the job completes
+func (j *Job) Done() <-chan struct{} {
+	return j.done
+}
+
+// setStatus sets the job status. status can only move to higher values:
+// not_started -> running -> complete -> stopped
+func (j *Job) setStatus(st Status) {
+	for {
```

Comment:
Is this for loop necessary? Looking at the implementation and operation ordering, I wonder if store alone solves all the issues?

Reply:
The concern is if the job is manually stopped at the same moment that it completes. If the job goes first from `StatusRunning` then to `StatusCompleted` in between the time that the `setStatus(StatusStopped)` gets the current status (as `StatusRunning`) and then does the `CompareAndSwap`, which fails because it is now `StatusCompleted`. The state should rightfully be changed to `StatusStopped` as the "higher" value, but without the loop, it won't be.

Reply:
After further review, I agree with both your assessment. StatusCompleted is set just before the `done` channel is closed and `Stop()` waits until after the `done` channel is closed to set the status. As such, there is no possibility of a race. I'll update. Thanks for the sharp eyes.

Reply:
And with that, status doesn't even need to be atomic either.